Playlists & crates: restore previous search

ronso0 opened the pull request:
Previous search queries are restored when selecting a playlist or crate.

To make users aware of this new feature we should finally highlight the searchbar when it's active #7922

Closes #11015 (see also #9522)

ronso0 committed:
Playlists (internal & external) and crates: restore previous search phrase

ronso0 commented on src/library/playlisttablemodel.h:
```diff
@@ -40,4 +40,5 @@ class PlaylistTableModel final : public TrackSetTableModel {
 
     int m_iPlaylistId;
     bool m_keepDeletedTracks;
+    QHash<int, QString> m_searchTexts;
```
I guess this could be moved to [TrackSetmodel](**https://github.com/mixxxdj/mixxx/blob/main/src/library/trackset/tracksettablemodel.h**) but I didn't dare to touch that minimal class..

daschuer commented on src/library/playlisttablemodel.cpp:
```diff
@@ -170,7 +175,16 @@ void PlaylistTableModel::setTableModel(int playlistId) {
-    setSearch("");
+
+    // Restore search text
+    QString prevSearch("");
```
QString is empty by default. 
```suggestion
    QString prevSearch;
```

daschuer commented on src/library/playlisttablemodel.cpp:
```diff
@@ -170,7 +175,16 @@ void PlaylistTableModel::setTableModel(int playlistId) {
-    setSearch("");
+
+    // Restore search text
+    QString prevSearch("");
+    if (m_iPlaylistId != -1) {
+        auto it = m_searchTexts.constFind(m_iPlaylistId);
+        if (it != m_searchTexts.constEnd()) {
+            prevSearch = it.value();
+        }
+    }
+    setSearch(prevSearch);
```
I think we can simplify this even more, because m_iPlaylistId will never be -1 and even if, the QHash will return an empty QString: 
```suggestion
    setSearch(m_searchTexts.value(m_iPlaylistId));
```

ronso0 replied:
I tried that first but it's not working because in the called code the string is tested for `isNull` (not `isEmty`).
I can check whether `isEmpty` can be used down the line.

ronso0 replied:
QString() didn't work because of
https://github.com/mixxxdj/mixxx/blob/3aa2318e34f306808420a6707857dabd37d0bea2/src/widget/wsearchlineedit.cpp#L456-L464

oh, this was you, eight years ago : ) 
https://github.com/mixxxdj/mixxx/commit/54f71ea7b8e5d6e87ad9132bcf7fdfe2a7c8372a#diff-5112567721f7d2345d3df3bdf1e73badf29f2feb7796f069777b54b949c910feL120

I'll remove the isNull branch. It is not required anymore since `disableSearch` is called explicitly where required.

ronso0 replied:
done.

daschuer reviewed (changes requested):
Some ideas for improvement

daschuer commented:
Is this a 2.3 fix?

ronso0 committed:
WSearchLineEdit: don't allow slotRestoreSearch to disable the searchbar

This is not needed anymore and unintenionally disables the searchbar when calling
BaseSqlTableModel::setSearch with search text QString() instead of QString("")

ronso0 committed:
external libraries: disable search in root view

daschuer reviewed (approved):
LGTM, Thank you.
